s3url/s3url.go

```diff
@@ -0,0 +1,195 @@
+	wildOperation := loc > -1
+	if !wildOperation {
+		s.Delimiter = s3Separator
+		s.Prefix = s.Key
+	} else {
+		s.Prefix = s.Key[:loc]
+		s.filter = s.Key[loc:]
+	}
+
+	filterRegex := matchAllRe
+	if s.filter != "" {
+		filterRegex = regexp.QuoteMeta(s.filter)
+		filterRegex = strings.Replace(filterRegex, "\\?", ".", -1)
+		filterRegex = strings.Replace(filterRegex, "\\*", ".*?", -1)
+	}
+	filterRegex = s.Prefix + filterRegex
+	r, err := regexp.Compile("^" + filterRegex + "$")
+	if err != nil {
+		return err
+	}
+	s.filterRegex = r
+	return nil
+}
+
+// Clone creates a new s3url with the values from the receiver.
+func (s S3Url) Clone() S3Url {
+	return S3Url{
+		Bucket:      s.Bucket,
+		Delimiter:   s.Delimiter,
+		Key:         s.Key,
+		Prefix:      s.Prefix,
+		filter:      s.filter,
+		filterRegex: s.filterRegex,
+	}
+}
+
+// Match check if given key matches with regex and
+// returns parsed key.
+func (s S3Url) Match(key string) (string, bool) {
+	if !s.filterRegex.MatchString(key) {
+		return "", false
+	}
+
+	isBatch := s.filter != ""
+	if isBatch {
+		return parseBatch(s.Prefix, key), true
+	}
+	return parseNonBatch(s.Prefix, key), true
+}
+
+// parseBatch parses keys for wildcard operations.
+// It cuts the key starting from first directory before the
+// wildcard part (filter)
+//
+// Example:
+//		key: a/b/test2/c/example_file.tsv
+//		prefix: a/b/
+//		output: test2/c/example_file.tsv
+//
+func parseBatch(prefix string, key string) string {
+	index := strings.LastIndex(prefix, s3Separator)
+	if index < 0 || !strings.HasPrefix(key, prefix) {
+		return key
+	}
+	trimmedKey := key[index:]
+	trimmedKey = strings.TrimPrefix(trimmedKey, s3Separator)
+	return trimmedKey
+}
+
+// parseNonBatch parses keys for non-wildcard operations.
+// It substracts prefix part from the key and gets first
+// path.
+//
+// Example:
+//		key: a/b/c/d
+//		prefix: a/b
+//		output: c/
+//
+func parseNonBatch(prefix string, key string) string {
+	if key == prefix || !strings.HasPrefix(key, prefix) {
+		return key
+	}
+	parsedKey := strings.TrimPrefix(key, prefix)
+	parsedKey = strings.TrimPrefix(parsedKey, s3Separator)
+	index := strings.Index(parsedKey, s3Separator) + 1
+	if index <= 0 || index >= len(parsedKey) {
+		return parsedKey
+	}
+	trimmedKey := parsedKey[:index]
+	return trimmedKey
+}
+
+// HasWild checks if a string contains any S3 wildcard chars.
+func HasWild(s string) bool {
+	return strings.ContainsAny(s, s3WildCharacters)
+}
+
+// New creates new S3Url by parsing given url.
+func New(rawUrl string) (*S3Url, error) {
+	if !strings.HasPrefix(rawUrl, s3Prefix) {
+		return nil, fmt.Errorf("s3 url should start with %s", s3Prefix)
+	}
+	parts := strings.SplitN(rawUrl, s3Separator, 4)
+	if parts[2] == "" {
+		return nil, fmt.Errorf("s3 url should have a bucket")
+	}
+	if HasWild(parts[2]) {
+		return nil, fmt.Errorf("bucket name cannot contain wildcards")
+	}
+	key := ""
+	if len(parts) == 4 {
+		key = strings.TrimLeft(parts[3], s3Separator)
```

Comment:
is this TrimLeft or TrimPrefix?

Reply:
`TrimLeft` would be a better choice since it removes all leading "/" characters while TrimPrefix does not. 

Example: Trimming `s3://bucket////key/b/c` should result with `key/b/c`. But if we use TrimPrefix, only the leading `/` will be removed and output will be like "//key/b/c". 
See: https://play.golang.org/p/GRP3JeKbl-5

Reply:
👍